File: Programmers/Level3/dfs/단어변환.cpp
#include <string>
#include <vector>

using namespace std;

int answer=0;

bool check(string a,string b){
    int idx=0;
    int cnt=0;
   
    while(idx<a.size()){
        if(a[idx]==b[idx]){
            cnt++;
        }
        idx++;
    }
    if(cnt==a.size()-1){
        return true;
    }
    return false;
}


void dfs(string begin,string target,vector<string>&words,int cnt){
    
    if(check(begin,target)){
        answer++;
        return;
    }
      if(cnt>words.size()){
        answer=0;
        return;
    }
    
    else{
        if(check(begin,words[cnt])){
            answer++;    
            dfs(words[cnt],target,words,cnt+1);
        }
        else{
            dfs(begin,target,words,cnt+1);
        }
    }
    
    
}


int solution(string begin, string target, vector<string> words) {
    
    
    bool flag=true;
    for(int i=0;i<words.size();i++){
        if(words[i]==target){
            flag=false;
        }
    }
    
    if(!flag){
        dfs(begin,target,words,0);
        
    }
    else{
        answer=0;
    }
    
    
    
    
    
    
    
    return answer;
}

File: Programmers/Level3/dfs/단어변환.py
answer = 9999

def check_answer(cur_word,tar_word):
    cnt=0
    for i in range(len(cur_word)):
        if cur_word[i]!=tar_word[i]:
            cnt+=1

    if cnt==1:
        return True

    return False

def dfs(start,target,words,visited,depth):
    global answer

    if start==target:
        answer=min(answer,depth)
        return

    for i in range(len(words)):
        if visited[i]==False and check_answer(start,words[i]):
           visited[i]=True
           dfs(words[i],target,words,visited,depth+1)
           visited[i]=False



    return 

def solution(begin, target, words):

    visited=[False]*len(words)
    dfs(begin,target,words,visited,0)

    if answer==9999:
        return 0

    return answer

